### DISCUSSION TITLE
Diagram of needle structure in a volume file

### DISCUSSION THREAD

**Discussion by bingoohuang** [seaweedfs/seaweedfs#1822] [Ideas]
![image](https://user-images.githubusercontent.com/1940588/108492942-b29ff400-72e0-11eb-98a9-a07eb8bc7d14.png)

[海草相关.pptx](https://github.com/chrislusf/seaweedfs/files/6009268/default.pptx)

@chrislusf source slide attached here!

I customized a Version4 format to support:

1. Change padding algorithm from (0,8] to [0,8), which means when payload size is already 8 bytes padded, padding bytes will be zero instead of 8 in version1,2,3.
2. To support iterate backwards for customized storage checking in random.

May i create a PR for the Version4 format?

**Comment by chrislusf** [comment]
I prefer not. The version change is not trivial and will impact many users. The new version has some benefits, but not large enough.

1. this saves 1 byte for 12.5% of files. 
2. you can traverse randomly with a idx file.

**Comment by kmlebedev** [comment]
It is not clear whether this improves reliability ?

**Reply by bingoohuang** [reply]
Of course, this at lease help me understand under-structure of a needle other than so many code line.

**Comment by dch** [comment]
this diagram is really helpful. Can you post it somewhere, maybe wiki, with a link to relevant source code, just the "current state" not the proposed change?
